server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/DKIMSign.java

```diff
@@ -0,0 +1,182 @@
+		return signatureTemplate;
+	}
+
+	/**
+	 * @return the privateKey
+	 */
+	protected PrivateKey getPrivateKey() {
+		return privateKey;
+	}
+
+    public void init() throws MessagingException {
+        signatureTemplate = getInitParameter("signatureTemplate");
+        String privateKeyString = getInitParameter("privateKey");
+        String privateKeyPassword = getInitParameter("privateKeyPassword", null);
+        forceCRLF = getInitParameter("forceCRLF", true);
+        try {
+            PKCS8Key pkcs8 = new PKCS8Key(new ByteArrayInputStream(
+                    privateKeyString.getBytes()),
+                    privateKeyPassword != null ? privateKeyPassword
+                            .toCharArray() : null);
+            privateKey = pkcs8.getPrivateKey();
+            // privateKey = DKIMSigner.getPrivateKey(privateKeyString);
+        } catch (NoSuchAlgorithmException e) {
+            throw new MessagingException("Unknown private key algorythm: "
+                    + e.getMessage(), e);
+        } catch (InvalidKeySpecException e) {
+            throw new MessagingException(
+                    "PrivateKey should be in base64 encoded PKCS8 (der) format: "
+                            + e.getMessage(), e);
+        } catch (GeneralSecurityException e) {
+            throw new MessagingException("General security exception: "
+                    + e.getMessage(), e);
+        }
+    }
+
+    public void service(Mail mail) throws MessagingException {
+        DKIMSigner signer = new DKIMSigner(getSignatureTemplate(), getPrivateKey());
+        SignatureRecord signRecord = signer
+                .newSignatureRecordTemplate(getSignatureTemplate());
+        try {
+            BodyHasher bhj = signer.newBodyHasher(signRecord);
+            MimeMessage message = mail.getMessage();
+            Headers headers = new MimeMessageHeaders(message);
+            try {
+            	OutputStream os = new HeaderSkippingOutputStream(bhj.getOutputStream());
+            	if (forceCRLF) os = new CRLFOutputStream(os);
+                message.writeTo(os);
+                bhj.getOutputStream().close();
+            } catch (IOException e) {
+                throw new MessagingException("Exception calculating bodyhash: "
+                        + e.getMessage(), e);
+            }
+            String signatureHeader = signer.sign(headers, bhj);
+            // Unfortunately JavaMail does not give us a method to add headers
+            // on top.
+            // message.addHeaderLine(signatureHeader);
+            prependHeader(message, signatureHeader);
+        } catch (PermFailException e) {
+            throw new MessagingException("PermFail while signing: "
+                    + e.getMessage(), e);
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    private void prependHeader(MimeMessage message, String signatureHeader)
+            throws MessagingException {
+        List<String> prevHeader = new LinkedList<String>();
```

Comment:
Moving is one issue, refactoring is an other. Do we want to take time doing this?

(It is an external contribution, we can ask Shrumit to do it...)